### ISSUE TITLE
Elasticsearch v5 Error WIth Cuckoov2-RC2?

### ISSUE THREAD

**Issue by keithjjones** [cuckoosandbox/cuckoo#1178]
I have a fresh install of Cuckoo v2-RC2 on a mac working really well.  I tried to tie it to Elasticsearch through brew (v5 is now the current version there).  I'm getting these errors and no data is being saved to ES.  Any ideas?  Is this a bug?

```
2016-11-21 14:28:13,279 [elasticsearch] WARNING: POST /cuckoov2-1-2016/cuckoo?op_type=create [status:400 request:0.026s]
2016-11-21 14:28:13,294 [lib.cuckoo.core.plugins] WARNING: The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #1: TransportError(400, u'action_request_validation_exception', u'Validation Failed: 1: an id must be provided if version type or value are set;')
```

That's all it says in cuckoo.log.

This is the relevant reporting.conf information I am using:

```
[elasticsearch]
enabled = yes
# Comma-separated list of ElasticSearch hosts. Format is IP:PORT, if port is
# missing the default port is used.
# Example: hosts = 127.0.0.1:9200, 192.168.1.1:80
hosts = 127.0.0.1:9200
# Set to yes if we want to be able to search every API call instead of just
# through the behavioral summary.
calls = no
# Index of this Cuckoo instance. If multiple Cuckoo instances connect to the
# same ElasticSearch host then this index (in Moloch called "instance") should
# be unique for each Cuckoo instance.
#
index = cuckoov2-1
#
# Logging time pattern.  This sets how elasticsearch creates indexes
# by default it is yearly in most instances this will be sufficient
# valid options: yearly, monthly, daily
#
index_time_pattern = yearly
#
# Cuckoo node name in Elasticsearch to identify reporting host. Can be useful
# for automation and while referring back to correct Cuckoo host. 
#
# cuckoo_node = "cuckoo.example.host"
```

**Comment by keithjjones** [comment]
Well, it's not ElasticSearch on a mac that is causing an issue.  I stood up an ElasticSearch instance on a CentOS computer and I'm getting the same type of error:

```
2016-11-21 16:09:13,861 [elasticsearch] WARNING: HEAD /_template/cuckoo_template [status:404 request:0.016s]
2016-11-21 16:09:15,294 [elasticsearch] WARNING: POST /cuckoov2-1-2016/cuckoo?op_type=create [status:400 request:0.771s]
2016-11-21 16:09:15,311 [lib.cuckoo.core.plugins] WARNING: The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #1: TransportError(400, u'action_request_validation_exception', u'Validation Failed: 1: an id must be provided if version type or value are set;')
```

**Comment by hoppers2k9** [comment]
It's an elastic 5 compatibility issue. It was only properly released about a week ago. Cuckoo uses an elastic method called create, which in the new version requires an ID to be provided, in elastic 5 the index method should be used. It's not a big fix but I presume it would need an updated dependency and would remove backwards compatibility.

**Comment by keithjjones** [comment]
Thanks @hoppers2k9 that is good to know.  A fresh install of ES 2.4.1 produces this error message.  ES 5 never touched that machine (I pulled back a VM snapshot before I touched it with ES 5).

```
016-11-21 16:36:55,671 [elasticsearch] WARNING: HEAD /_template/cuckoo_template [status:404 request:0.014s]
2016-11-21 16:36:58,950 [elasticsearch] WARNING: POST /cuckoov2-1-2016/cuckoo?op_type=create [status:400 request:2.591s]
2016-11-21 16:36:58,967 [lib.cuckoo.core.plugins] WARNING: The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #1: TransportError(400, u'mapper_parsing_exception', u'failed to parse')
```

Any ideas?

Also, I imagine you probably want this issue to stay active for v5 until there is a solution?

Thanks!

**Comment by razuz** [comment]
hmm ... that's interesting ... I'm running on 2.4.1 and do not have this issue. However I'll double check that tomorrow just in case.

just to check - will 2nd analyze work OK or generate the same thing ?

**Comment by keithjjones** [comment]
It generates the same error on subsequent runs @razuz

**Comment by keithjjones** [comment]
@razuz Are you using requirements.txt?  That specifies Python module for ES 2.2.0.  I tested with Python client 2.4.0 (didn't have 2.4.1) and ES 2.4.1 and it didn't work.

**Comment by keithjjones** [comment]
ES v2.2.0 and python module v2.2.0 gives the same error.  I'm out of ideas.

**Comment by keithjjones** [comment]
Ok, I got it working with python elasticsearch module 2.4.0 and ES 2.4.1 on a mac.  You have to get the versions perfect on both sides.  I still get this warning:

```
2016-11-21 19:07:24,651 [elasticsearch] WARNING: Undecodable raw error response from server: No JSON object could be decoded
```

I don't know what that is about, but it seems to search the fields like I think it's supposed to.  Do you want to keep this issue open for the ES v5 compatibility later?

**Comment by keithjjones** [comment]
Another change I made was taking any dashes out of the "cuckoo" index name within reporting.conf.  That may have caused problems too.

**Comment by keithjjones** [comment]
Verified it has something to do with the sample:

http://www.malware-traffic-analysis.net/2016/11/15/index.html
File named:  2016-11-15-3rd-run-Rig-V-payload-Cerber.exe

For some reason it is reporting something that Elasticsearch doesn't like when submitted by cuckoo.  Had I tried a different sample all this time, I would have seen results.  This one messes up ES every time.  Is there some way to, at a minimum, input hashes in ES if something else failed so you can still search for hashes with the search box?  It seems to be an "all or nothing" shot with ES and if it fails it won't be searched by the search box at all.  It is likely if this one fails there will be others that fail.

Is there a way to cross check what is in the database with ElasticSearch and report which samples are not in both?

**Comment by keithjjones** [comment]
Confirmed that this is happening on other samples as well.  I'm not sure what in a particular sample is causing this error, but the data cannot be found using the search box if it happens.

```
2016-11-22 11:35:41,215 [elasticsearch] WARNING: POST /cuckoo-2016/cuckoo?op_type=create [status:400 request:0.004s]
2016-11-22 11:35:41,216 [lib.cuckoo.core.plugins] WARNING: The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #2: TransportError(400, u'mapper_parsing_exception', u'object mapping for [signatures.marks.section] tried to parse field [section] as object, but found a concrete value')
```

**Comment by keithjjones** [comment]
Elasticsearch 2.4 is still causing errors with Cuckoo v2:

```
WARNING:elasticsearch:POST http://127.0.0.1:9200/cuckoo-2016/cuckoo?op_type=create [status:N/A request:10.055s]
Traceback (most recent call last):
  File "/Source/Python2/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 94, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/Source/Python2/lib/python2.7/site-packages/urllib3/connectionpool.py", line 643, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/Source/Python2/lib/python2.7/site-packages/urllib3/util/retry.py", line 311, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/Source/Python2/lib/python2.7/site-packages/urllib3/connectionpool.py", line 594, in urlopen
    chunked=chunked)
  File "/Source/Python2/lib/python2.7/site-packages/urllib3/connectionpool.py", line 393, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/Source/Python2/lib/python2.7/site-packages/urllib3/connectionpool.py", line 313, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
ReadTimeoutError: HTTPConnectionPool(host=u'127.0.0.1', port=9200): Read timed out. (read timeout=10)
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #5: ConnectionTimeout caused by - ReadTimeoutError(HTTPConnectionPool(host=u'127.0.0.1', port=9200): Read timed out. (read timeout=10))
ERROR:lib.cuckoo.core.plugins:Failed to run the reporting module "MongoDB":
Traceback (most recent call last):
  File "/Source/cuckoo-dev/utils/../lib/cuckoo/core/plugins.py", line 533, in process
    current.run(self.results)
  File "/Source/cuckoo-dev/utils/../modules/reporting/mongodb.py", line 253, in run
    self.db.analysis.save(report)
  File "/Source/Python2/lib/python2.7/site-packages/pymongo/collection.py", line 1903, in save
    check_keys, manipulate, write_concern)
  File "/Source/Python2/lib/python2.7/site-packages/pymongo/collection.py", line 430, in _insert
    gen(), check_keys, self.codec_options, sock_info)
DocumentTooLarge: BSON document too large (17230078 bytes) - the connected server supports BSON document sizes up to 16777216 bytes.
```

The mongodb error is something I discussed in Issue #358

**Comment by jbremer** [comment]
Sorry - I'm only seeing this issue just now. Regarding the last two stacktraces, could you please share the entire analysis?

**Comment by keithjjones** [comment]
Here is some more.  I --cleaned and tried again.  Then I cleaned out the ES index with Curl and restarted the daemon.

```
INFO:cuckoo.process:Task #5: queueing for reporting
WARNING:elasticsearch:POST /cuckoo-2016/cuckoo?op_type=create [status:400 request:0.414s]
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #2: TransportError(400, u'mapper_parsing_exception', u'failed to parse')
ERROR:lib.cuckoo.core.plugins:Failed to run the reporting module "MongoDB":
Traceback (most recent call last):
  File "/Source/cuckoo-dev/utils/../lib/cuckoo/core/plugins.py", line 533, in process
    current.run(self.results)
  File "/Source/cuckoo-dev/utils/../modules/reporting/mongodb.py", line 253, in run
    self.db.analysis.save(report)
  File "/Source/Python2/lib/python2.7/site-packages/pymongo/collection.py", line 1903, in save
    check_keys, manipulate, write_concern)
  File "/Source/Python2/lib/python2.7/site-packages/pymongo/collection.py", line 430, in _insert
    gen(), check_keys, self.codec_options, sock_info)
DocumentTooLarge: BSON document too large (17399311 bytes) - the connected server supports BSON document sizes up to 16777216 bytes.
```

**Comment by keithjjones** [comment]
@jbremer What do you mean share the entire analysis?

**Comment by keithjjones** [comment]
Another error:

```
WARNING:elasticsearch:POST /cuckoo-2016/cuckoo?op_type=create [status:400 request:0.003s]
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #5: TransportError(400, u'mapper_parsing_exception', u'object mapping for [signatures.marks.section] tried to parse field [section] as object, but found a concrete value')
```

Signatures.marks.section?

**Comment by swackhamer** [comment]
Could you try: https://github.com/swackhamer/cuckoo-elasticsearch

Signatures has a issue in the current version of the reporting module in mainline but it is fixed in that release.  If this works for you I will merge the changes.  I have been meaning to open a PR for this issue.

The problem is with signatures outputting calls that are sometimes numbers and sometimes strings.

You will need to delete your Cuckoo indexes and your cuckoo index template.  A curl command should work for that

curl -XDELETE http://localhost:9200 /_template/cuckoo_template


The error you gave here:     raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)

is a problem with the elasticsearch server not having enough resources to index the document.  We can increase the request_timeout on the indexing call to fix this.

**Comment by swackhamer** [comment]
It looks like in ES 5.0 we need to explicitly state the op type to "index" and not the implicit "create".  That should solve your first issue.

https://github.com/swackhamer/cuckoo-elasticsearch/blob/master/modules/reporting/elasticsearch.py#L190

We have to add the line there.  I will fix it in my module and then submit the PR after you ensure that the following fix also works for you.

"_op_type": "index"

**Comment by keithjjones** [comment]
I'm happy to try it @swackhamer.  What do I do with that repository to get it to work in my cuckoo?  Do I just copy your repository over top my cuckoo directory?

**Comment by swackhamer** [comment]
Replace template.json with https://github.com/swackhamer/cuckoo-elasticsearch/blob/master/data/elasticsearch/template.json in data/elasticsearch and replace elasticsearch.py in modules/reporting/ with https://github.com/swackhamer/cuckoo-elasticsearch/blob/master/modules/reporting/elasticsearch.py

I also added in the 5.0 "fixes".  Let me know if that fixes the issue on 5.0 as well.

Make sure to delete your cuckoo indexes AND your cuckoo_template before you rerun Cuckoo with these changes.

**Comment by swackhamer** [comment]
There was another explicit call to create in there that I just fixed, make sure to get the most recent version of elasticsearch.py for ES 5.0 compatibility.

**Comment by keithjjones** [comment]
Testing now.  I'm testing with ES 2.4 first, then I'll install ES 5 and give it a whirl.

**Comment by keithjjones** [comment]
@swackhamer I got the following error during processing:

```
INFO:cuckoo.process:Task #5: queueing for reporting
WARNING:elasticsearch:POST /cuckoo-2016/cuckoo [status:400 request:0.651s]
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #2: TransportError(400, u'mapper_parsing_exception', u'failed to parse')
```

**Comment by swackhamer** [comment]
Can you compress your archived report in storage/analysis and send me the file?

I have seen that mongodb problem specifically on files that create huge API reports like Office Documents.  I can't help you with that but I should be able to get this ES report working.  Unfortunately the error is too vague and I need the full analyses report.  There will be some information about your sandbox in the report so you may not want to post it publicly, but if you could send me a email with the archive I will try to get it processed now: allen [.] swackhamer at gmail .

tar cvf analyses.tar storage/analyses/4/

**Comment by keithjjones** [comment]
Sure, no problem.  You said task 4, but it looks like task 2 was the one that failed.  You want the one that failed, correct?

**Comment by swackhamer** [comment]
Yep task #2 sorry.

**Comment by razuz** [comment]
one of our boxes got everything fixed by just swapping `self.es.create` to `self.es.index` on line https://github.com/cuckoosandbox/cuckoo/blob/master/modules/reporting/elasticsearch.py#L127
as also referred in @swackhamer fix 

sadly cannot dig deeper right now

**Comment by keithjjones** [comment]
It's huge.  I will send you a google drive link.  Let me know when you get it @swackhamer

**Comment by keithjjones** [comment]
Verified that @swackhamer's fix works for ES 5.0.2 on a Mac, @razuz

**Comment by razuz** [comment]
cool :) nice work ! :)
I'll do that in couple weeks tho' :/

**Comment by swackhamer** [comment]
I am working through the report that keithjjones sent me.  It produces a 1.4GB report.json file so this thing is massive.

Once I get this bug fixed I am going to submit a PR with these changes to support ES 5.0 compatibility.  I am also testing with the new elasticsearch 5.0 python module which will work for both 2.X and 5.X.

**Comment by zippy1990** [comment]
@razuz Can confirm that this works for ES 5.0 but it still produces the same errors that you would ocasionnaly get on ES 2.4.2, namely:

`WARNING:elasticsearch:POST /cuckoo-2016/cuckoo [status:400 request:0.014s]
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #3: TransportError(400, u'illegal_argument_exception', u'[ZQliz-i][127.0.0.1:9300][indices:data/write/index[p]]')
`

and

`ERROR:sqlalchemy.pool.QueuePool:Exception during reset or similar
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/pool.py", line 636, in _finalize_fairy
    fairy._reset(pool)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/pool.py", line 774, in _reset
    self._reset_agent.rollback()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1563, in rollback
    self._do_rollback()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1601, in _do_rollback
    self.connection._rollback_impl()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 670, in _rollback_impl
    self._handle_dbapi_exception(e, None, None, None, None)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 668, in _rollback_impl
    self.engine.dialect.do_rollback(self.connection)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py", line 420, in do_rollback
    dbapi_connection.rollback()
DatabaseError: (psycopg2.DatabaseError) error with status PGRES_TUPLES_OK and no message from the libpq`


It's strange because both seem to happen at random once in 10 reports or so.

**Comment by keithjjones** [comment]
Working on a fix in Pull #1195

**Comment by jbremer** [comment]
Well, as we merged #1196 a while ago, I suppose this issue may be closed? Please reopen or create a new one if you run into new or related issues with the latest version(s) of Cuckoo! Thanks.

**Comment by SparkyNZL** [comment]
@jbremer you just need to merge the change for the creation of the ES
cuckoo index to allow for more than 1000 results ;)
